### PR TITLE
`MessagesGet`

### PR DESCRIPTION
**Pull request by mistermoe**
Final PR to round out sync foundations. preceded by #305 

Introduces an `MessagesGet` message to retrieve messages.  Will include more details in this description tmrw. brain is total mush right now.

### CODE REVIEW

File: src/interfaces/messages/handlers/messages-get.ts [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    const messages = await Promise.all(promises);
+
+    // for every message, include associated data as `encodedData` IF:
+    //  * its a RecordsWrite
+    //  * the data size is equal or smaller than the size threshold
+    //! NOTE: this is somewhat duplicate code that also exists in `StorageController.query`.
+    //! reviewers: consider whether we oughta abstract the common aspects.
+    for (const entry of messages) {
```
Comment (mistermoe) [inline src/interfaces/messages/handlers/messages-get.ts:58]: will refactor to use `Promise.all`

File: src/interfaces/messages/handlers/messages-get.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
+      const promise = this.messageStore.get(tenant, messageCid)
+        .then(message => {
+          return { messageCid, message };
+        })
+        .catch(_ => {
+          // FIXIT: log error? throw error?
+          return { messageCid, message: undefined };
```
Comment (diehuxx) [inline src/interfaces/messages/handlers/messages-get.ts:45]: Do we have any error logging set up for DWNS? I'm not aware of it, but seems like a reasonable action to take for this error.

I don't mind leaving `message` empty, but also could dig adding an `error` field to `MessagesGetReplyEntry`. Throwing seems like like overkill.

File: src/interfaces/messages/handlers/messages-get.ts [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+      promises.push(promise);
+    }
+
+    const messages = await Promise.all(promises);
+
+    // for every message, include associated data as `encodedData` IF:
+    //  * its a RecordsWrite
+    //  * the data size is equal or smaller than the size threshold
+    //! NOTE: this is somewhat duplicate code that also exists in `StorageController.query`.
+    //! reviewers: consider whether we oughta abstract the common aspects.
```
Comment (diehuxx) [inline src/interfaces/messages/handlers/messages-get.ts:57]: I don't think this duplication is egregious that it _needs_ to be abstracted, but personally I would make a `StorageController#addEncodedData(message: BaseMessage): BaseMessage)` (or maybe call it `conditionallyAddEncodedData` idk) method that contains lines 65-81. Checks if should add `encodedData`, then adds it.

File: src/interfaces/messages/messages/messages-get.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+    const message = { descriptor, authorization };
+
+    Message.validateJsonSchema(message);
+    this.validateMessageCids(options.messageCids);
+
+    return new MessagesGet(message);
+  }
+
+  /**
+   * Q for reviewers: worth it?
```
Comment (diehuxx) [inline src/interfaces/messages/messages/messages-get.ts:40]: Eh, might as well. Seems harmless. I might even go so far as to add a custom `format` to the JSON schema.

File: src/interfaces/messages/types.ts [hunk 5]
```diff
@@ -0,0 +1,24 @@
+import type { BaseMessageReply } from '../../core/message-reply.js';
+import type { BaseMessage, Descriptor } from '../../core/types.js';
+import type { DwnInterfaceName, DwnMethodName } from '../../core/message.js';
+
+export type MessagesGetDescriptor = Pick<Descriptor, 'interface' | 'method' > & {
```
Comment (diehuxx) [inline src/interfaces/messages/types.ts:5]: Just thinking out loud: I don't like `Pick`s very much. Would prefer to reduce the fields in `Descriptor` and move non-universal fields into message-specific descriptor types. But that is outside the scope of this PR.

Comment (mistermoe) [reply src/interfaces/messages/types.ts:5]: I agree! Gave it a shot because it was mentioned in the last PR.

File: src/interfaces/messages/handlers/messages-get.ts (hunk 2 above)
Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts:45]: Re: adding `error` field

great idea @diehuxx! Will add today

File: src/interfaces/messages/types.ts [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+  method: DwnMethodName.Get;
+  messageCids: string[];
+};
+
+export type MessagesGetMessage = BaseMessage & {
+  descriptor: MessagesGetDescriptor;
+};
+
+// oof horrible name
+export type MessagesGetReplyEntry = {
```
Comment (diehuxx) [inline src/interfaces/messages/types.ts:16]: Another idea for a refactor: I've never used [typescript namespaces](https://www.typescriptlang.org/docs/handbook/namespaces.html#namespaced-validators) before, but it might make these type names less ridiculous.

Like so:
```
namespace MessagesGet {
  export type ReplyEntry = {
     ...
  };

  export type Message {
     ...
  }
}
```

Again this idea is outside the scope of this PR, so I'm fine with this goofy variable name for now.

File: tests/interfaces/messages/handlers/messages-get.spec.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    const alice = await DidKeyResolver.generate();
+    const { recordsWrite } = await TestDataGenerator.generateRecordsWrite({ requester: alice });
+    const recordsWriteMessageCid = await Message.getCid(recordsWrite.message);
+
+    const { message } = await TestDataGenerator.generateMessagesGet({
+      requester   : alice,
+      messageCids : [recordsWriteMessageCid]
+    });
+
+    // 0 messages expected because the RecordsWrite created above was never stored
```
Comment (diehuxx) [inline tests/interfaces/messages/handlers/messages-get.spec.ts:173]: wait, where does bob enter into this?

File: tests/interfaces/messages/handlers/messages-get.spec.ts [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+    const reply = await dwn.processMessage(alice.did, recordsWrite.toJSON(), dataStream);
+    expect(reply.status.code).to.equal(202);
+
+    const recordsWriteMessageCid = await Message.getCid(recordsWrite.message);
+    const { message } = await TestDataGenerator.generateMessagesGet({
+      requester   : alice,
+      messageCids : [recordsWriteMessageCid]
+    });
+
+    // 1 messages expected because the RecordsWrite created above is not bob's
```
Comment (diehuxx) [inline tests/interfaces/messages/handlers/messages-get.spec.ts:200]: wait, where does bob enter into this?

Comment (diehuxx) [reply tests/interfaces/messages/handlers/messages-get.spec.ts:200]: oh I see a similar comment in the test below which does have a bob persona. I might just omit this comment

File: tests/interfaces/messages/messages/messages-get.spec.ts [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+      try {
+        await MessagesGet.create({
+          authorizationSignatureInput : await Jws.createSignatureInput(alice),
+          messageCids                 : []
+        });
+
+        expect.fail();
+      } catch (e: any) {
+        expect(e.message).to.include('messageCids');
```
Comment (diehuxx) [inline tests/interfaces/messages/messages/messages-get.spec.ts:38]: nit: include more of the expected error message here. Makes it easier to CMD+f what I broke if I ever break this test one day.

File: tests/interfaces/messages/messages/messages-get.spec.ts [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+
+      try {
+        await MessagesGet.create({
+          authorizationSignatureInput : await Jws.createSignatureInput(alice),
+          messageCids                 : ['abcd']
+        });
+
+        expect.fail();
+      } catch (e: any) {
+        expect(e.message).to.include('valid CID');
```
Comment (diehuxx) [inline tests/interfaces/messages/messages/messages-get.spec.ts:53]: Same as above. More of error message pls.

File: tests/interfaces/messages/messages/messages-get.spec.ts [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+      }
+    });
+  });
+
+  describe('parse', () => {
+    it('parses a message into an MessagesGet instance', async () => {
```
Comment (diehuxx) [inline tests/interfaces/messages/messages/messages-get.spec.ts:59]:
```suggestion
    it('parses a message into a MessagesGet instance', async () => {
```

**Review by diehuxx — COMMENTED** [review]
looking solid! no major issues, just an assortment of small things.

File: tests/interfaces/messages/messages/messages-get.spec.ts [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+
+      const message = messagesGet.toJSON() as MessagesGetMessage;
+      message.descriptor.messageCids = ['abcd'];
+
+      try {
+        await MessagesGet.parse(message);
+
+        expect.fail();
+      } catch (e: any) {
+        expect(e.message).to.include('valid CID');
```
Comment (diehuxx) [inline tests/interfaces/messages/messages/messages-get.spec.ts:94]: Same as above. More of error message pls.

File: src/interfaces/messages/handlers/messages-get.ts [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    }
+
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
+      const promise = this.messageStore.get(tenant, messageCid)
+        .then(message => {
+          return { messageCid, message };
+        })
+        .catch(_ => {
```
Comment (dcrousso) [inline src/interfaces/messages/handlers/messages-get.ts:43]: i dont think `MessageStore` will ever throw (unless the operation is aborted).  you probably should check for `message === undefined` instead

File: src/interfaces/messages/handlers/messages-get.ts [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+          // FIXIT: log error? throw error?
+          return { messageCid, message: undefined };
+        });
+
+      promises.push(promise);
+    }
+
+    const messages = await Promise.all(promises);
```
Comment (dcrousso) [inline src/interfaces/messages/handlers/messages-get.ts:51]: NIT: instead of creating a `promises` array you could do this with `Array.prototype.map`

and also at the same time you could merge the `encodedData` part so that we dont iterate over all the `messages` twice

```
const messages = await Promise.all(message.descriptor.messageCids.map(async messageCid => {
  const message = await this.messageStore.get(tenant, messageCid);
  if (!message) {
    return { messageCid, error }; // FIXME
  }

  if (message.descriptor.interface === DwnInterfaceName.Records && message.descriptor.method === DwnMethodName.Write) {
    const dataCid = message.descriptor.dataCid;
    const dataSize = message.descriptor.dataSize;
    if (dataCid !== undefined && dataSize! <= DwnConstant.maxDataSizeAllowedToBeEncoded) {
      const messageCid = await Message.getCid(message);
      const { dataStream } = await dataStore.get(tenant, messageCid, dataCid) ?? { };
      if (dataStream) {
        const dataBytes = await DataStream.toBytes(dataStream);
        message.encodedData = Encoder.bytesToBase64Url(dataBytes);
      }
    }
  }

  return { messageCid, message };
});
```

File: src/interfaces/messages/messages/messages-get.ts [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+
+export type MessagesGetOptions = {
+  messageCids: string[];
+  authorizationSignatureInput: SignatureInput;
+};
+
+export class MessagesGet extends Message<MessagesGetMessage> {
+  public static async parse(message: MessagesGetMessage): Promise<MessagesGet> {
+    Message.validateJsonSchema(message);
+    this.validateMessageCids(message.descriptor.messageCids);
```
Comment (dcrousso) [inline src/interfaces/messages/messages/messages-get.ts:16]: NIT: i personally avoid using `this` in a `static` context as it can be confusing at a glance (i.e. only use `this` to represent an instance, not the `class`).  id suggest using `MessagesGet.validateMessageCids` instead

File: src/interfaces/messages/messages/messages-get.ts (hunk 4 above)
Comment (dcrousso) [reply src/interfaces/messages/messages/messages-get.ts:40]: i dont think we do this kind of validation anywhere else (e.g. `dataCid`) so i think it's probably fine to not do it

File: src/interfaces/records/handlers/records-delete.ts [hunk 16]
```diff
@@ -47,6 +47,7 @@ export class RecordsDeleteHandler implements MethodHandler {
     const newestExistingMessage = await RecordsWrite.getNewestMessage(existingMessages);
     let incomingMessageIsNewest = false;
     let newestMessage;
+
```
Comment (dcrousso) [inline src/interfaces/records/handlers/records-delete.ts:50]: oops?

File: src/interfaces/messages/handlers/messages-get.ts [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+    try {
+      await authenticate(message.authorization, this.didResolver);
+      await authorize(tenant, messagesGet);
+    } catch (e) {
+      return MessageReply.fromError(e, 401);
+    }
+
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+
+    for (const messageCid of message.descriptor.messageCids) {
```
Comment (dcrousso) [inline src/interfaces/messages/handlers/messages-get.ts:38]: should we attempt to do any deduplication here in case the caller provided the same `messageCid` more than once?

File: src/interfaces/messages/types.ts [hunk 18]
```diff
@@ -0,0 +1,24 @@
...
+  messageCid: string;
```
Comment (dcrousso) [inline src/interfaces/messages/types.ts:17]: i wonder if we should return a `Map`/`Object` where the keys are the `messageCid` and the values are the `{ message, encodedData }` so that the caller can:

- ensure they will never encounter the same `messageCid` more than once (since keys are unique)
- quickly grab the `message`/`encodedData` for a given `messageCid` instead of having to iterate the entire array to find the entry matching the `messageCid`
- possibly even avoid storing the `messageCid` (at least without having to manipulate the data) if they wanted to do something like `savedMessages.push(Object.values(result))`

**Review by dcrousso — DISMISSED** [review]

File: src/interfaces/messages/handlers/messages-get.ts (hunk 1 above)
Comment (michaelneale) [reply src/interfaces/messages/handlers/messages-get.ts:58]: a promise to use a promise

File: src/interfaces/messages/handlers/messages-get.ts (hunk 13 above)
Comment (diehuxx) [reply src/interfaces/messages/handlers/messages-get.ts:43]: `message-store-level` should not throw, but it's possible a custom message store will. In that case, it's useful to `catch`. Also, don't think we should throw/log error on `message === undefined`. That's a non-exceptional case IMO, and the client should be okay with receiving `message: undefined` in the reply entry.

File: src/interfaces/messages/messages/messages-get.ts (hunk 4 above)
Comment (mistermoe) [reply src/interfaces/messages/messages/messages-get.ts:40]: reason behind doing so was to prevent unnecessary i/o for bunk CIDs

File: src/interfaces/messages/handlers/messages-get.ts (hunk 17 above)
Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts]: good idea. done

File: src/interfaces/records/handlers/records-delete.ts (hunk 16 above)
Comment (mistermoe) [reply src/interfaces/records/handlers/records-delete.ts]: yeah instinctive fat pinkied. fixed

File: src/interfaces/messages/handlers/messages-get.ts (hunk 13 above)
Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts:44]: > `message-store-level` should not throw, but it's possible a custom message store will. In that case, it's useful to `catch`.  Also, don't think we should throw/log error on `message === undefined`. That's a non-exceptional case IMO, and the client should be okay with receiving `message: undefined` in the reply entry.

💯

File: tests/interfaces/messages/messages/messages-get.spec.ts (hunk 9 above)
Comment (mistermoe) [reply tests/interfaces/messages/messages/messages-get.spec.ts]: good call. this one specifically is auto-generated by AJV, our json schema parser/validator. still went ahead and included fuller error messages for all tests. In this case, hopefully the fuller error message will maaaybe help indicate that it's coming from json schema. went ahead and called that out in a comment

File: src/interfaces/messages/types.ts (hunk 18 above)
Comment (diehuxx) [reply src/interfaces/messages/types.ts:17]: > possibly even avoid storing the `messageCid` (at least without having to manipulate the data) if they wanted to do something like `savedMessages.push(Object.values(result))`

Not sure what you mean by this

> quickly grab the message/encodedData for a given messageCid instead of having to iterate the entire array to find the entry matching the messageCid

In practice, I think clients will generally be iterating over the array so they can process messages in order, but you're right that it may benefit the odd client. If we use an object with `messageCid`s as keys, we don't return an ordered list. [Looks like some JSON serializers don't preserve key order](https://filemakerhacks.com/2020/04/29/ordering-elements-within-json-objects/#:~:text=The%20JSON%20Data%20Interchange%20Standard,%2C%20and%20should%20not%2C%20matter.) That might not be a big deal; the client could remember the list of `messageCids` they sent in the request and use that to determine the order of messages to process. Not arguing against your approach, just something to think about as we put sync in production.

File: src/interfaces/messages/handlers/messages-get.ts (hunk 2 above)
Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts:46]: @diehuxx  for error message: `Failed to get message`?

File: src/interfaces/messages/messages/messages-get.ts (hunk 4 above)
Comment (mistermoe) [reply src/interfaces/messages/messages/messages-get.ts:40]: @dcrousso imo `dataCid` is validated when its integrity checked [here](https://github.com/TBD54566975/dwn-sdk-js/blob/main/src/store/storage-controller.ts#L70-L81).

Comment (mistermoe) [reply src/interfaces/messages/messages/messages-get.ts]: @diehuxx definitely considered adding a custom `format` to JSON schema but decided against it. Trying to keep use of json schema as language agnostic as possible in hopes that they can be used if anyone takes a stab at implementing any part of DWNs in another language

File: src/interfaces/messages/handlers/messages-get.ts (hunk 2 above)
Comment (diehuxx) [reply src/interfaces/messages/handlers/messages-get.ts:46]: @mistermoe worx 4 me

**Review by diehuxx — DISMISSED** [review]

File: src/interfaces/messages/handlers/messages-get.ts [hunk 19]
```diff
@@ -62,7 +63,9 @@ export class MessagesGetHandler implements MethodHandler {
       }
 
       const { interface: messageInterface, method } = message.descriptor;
-      if (messageInterface !== DwnInterfaceName.Records || method !== DwnMethodName.Write) {
+      const interfaceMethod = `${messageInterface}${method}`;
+
+      if (interfaceMethod !== RecordsWrite) {
```
Comment (dcrousso) [inline src/interfaces/messages/handlers/messages-get.ts]: im confused.  what was wrong with the old logic?

Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts]: just feel like it's easier to read

Comment (diehuxx) [reply src/interfaces/messages/handlers/messages-get.ts]: @mistermoe ngl disagree, but not gonna stop you.

having a var named `RecordsWrite` is confusing cuz dwn-sdk-js has a class named `RecordsWrite` already.
if we really wanna make this more readable, the ruby-ish thing that I would do is define a bunch of methods like `isRecordsWrite(message)`. might be overkill, but i tend to like that specific type of overkill.

Comment (dcrousso) [reply src/interfaces/messages/handlers/messages-get.ts]: ah.  personally i find this less easy to read, especially since there's already a `class RecordsWrite` elsewhere in this repo (tho we could have a `RecordsWrite.Identifier` or something?) 😕

perhaps we could flip this around and have it be `if (interface === DwnInterfaceName.Records && method === DwnMethodName.Write)` (or put a `if (!(...))` if you wanna keep it with `continue`)?

Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts]: the crowd has spoken! reverting

Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts]: what are y'all talking about? i didn't change anything 🥸

File: src/dwn.ts [hunk 20]
```diff
@@ -40,6 +41,7 @@ export class Dwn {
 
     this.methodHandlers = {
       [DwnInterfaceName.Events + DwnMethodName.Get]          : new EventsGetHandler(this.didResolver, this.eventLog),
+      [DwnInterfaceName.Messages + DwnMethodName.Get]        : new MessagesGetHandler(this.didResolver, this.messageStore, this.dataStore),
```
Comment (thehenrytsai) [inline src/dwn.ts:44]: Sanity: no explicit handler method? Was there more discussions on this while I was away?

File: src/interfaces/messages/types.ts (hunk 5 above)
Comment (thehenrytsai) [inline src/interfaces/messages/types.ts:5]: For my understanding: what is the intent of Picking `interface` and `method` when line 6 and 7 specifies it again?

File: src/interfaces/messages/handlers/messages-get.ts [hunk 21]
```diff
@@ -0,0 +1,88 @@
...
+    const promises: Promise<MessagesGetReplyEntry>[] = [];
+    const messageCids = new Set(message.descriptor.messageCids);
+
+    for (const messageCid of messageCids) {
+      const promise = this.messageStore.get(tenant, messageCid)
+        .then(message => {
+          return { messageCid, message };
+        })
+        .catch(_ => {
+          return { messageCid, message: undefined, error: `Failed to get message ${messageCid}` };
```
Comment (thehenrytsai) [inline src/interfaces/messages/handlers/messages-get.ts:45]: No test for this line.

File: tests/interfaces/messages/handlers/messages-get.spec.ts [hunk 22]
```diff
@@ -0,0 +1,240 @@
...
+      expect(messageReply.messageCid).to.not.be.undefined;
+      expect(messageReply.message).to.not.be.undefined;
+      expect(messageCids).to.include(messageReply.messageCid);
+
+      const cid = await Message.getCid(messageReply.message!);
+      expect(messageReply.messageCid).to.equal(cid);
+    }
+  });
+
+  it('returns message as undefined in reply entry when a messageCid is not found', async () => {
```
Comment (thehenrytsai) [inline tests/interfaces/messages/handlers/messages-get.spec.ts:163]: Sanity: is this meant to be the test that tests the line above that is never hit?

File: src/interfaces/messages/types.ts (hunk 5 above)
Comment (mistermoe) [reply src/interfaces/messages/types.ts]: removed `Pick` usage

File: src/dwn.ts (hunk 20 above)
Comment (mistermoe) [reply src/dwn.ts:45]: nothing to the contrary was discussed while you were gone. went ahead and added the explicit `handleMessagesGet` method to `dwn`

File: tests/interfaces/messages/handlers/messages-get.spec.ts (hunk 22 above)
Comment (mistermoe) [reply tests/interfaces/messages/handlers/messages-get.spec.ts:163]: na. this test is not supposed to hit the error code path

File: src/interfaces/messages/handlers/messages-get.ts (hunk 21 above)
Comment (mistermoe) [reply src/interfaces/messages/handlers/messages-get.ts:45]: added

**Review by thehenrytsai — APPROVED** [review]
Good to go! 🙇
